File: ums/frontend/fingerprint.js
/**
 * Browser Fingerprinting Utility for UMS
 *
 * This module provides client-side browser fingerprinting capabilities
 * for user recognition in the User Management Services.
 */

/**
 * Generate a comprehensive browser fingerprint
 * @returns {Promise<Object>} Fingerprint data object
 */
export async function generateFingerprint() {
    const fingerprint = {};

    try {
        // Basic browser information
        fingerprint.userAgent = navigator.userAgent;
        fingerprint.language = navigator.language;
        fingerprint.languages = navigator.languages ? Array.from(navigator.languages) : [];
        fingerprint.platform = navigator.platform;

        // Screen information
        fingerprint.screenResolution = [screen.width, screen.height];
        fingerprint.colorDepth = screen.colorDepth;
        fingerprint.pixelRatio = window.devicePixelRatio || 1;

        // Timezone
        fingerprint.timezone = Intl.DateTimeFormat().resolvedOptions().timeZone;

        // Hardware information (if available)
        if ('hardwareConcurrency' in navigator) {
            fingerprint.hardwareConcurrency = navigator.hardwareConcurrency;
        }

        if ('deviceMemory' in navigator) {
            fingerprint.deviceMemory = navigator.deviceMemory;
        }

        // Canvas fingerprinting
        fingerprint.canvas = await getCanvasFingerprint();

        // WebGL fingerprinting
        fingerprint.webgl = getWebGLFingerprint();

        // Font detection (simplified)
        fingerprint.fonts = await detectFonts();

        // Plugin information (if available)
        if (navigator.plugins && navigator.plugins.length > 0) {
            fingerprint.plugins = Array.from(navigator.plugins).map(plugin => plugin.name);
        }

        // Additional browser features
        fingerprint.features = {
            localStorage: typeof Storage !== 'undefined',
            sessionStorage: typeof sessionStorage !== 'undefined',
            indexedDB: typeof indexedDB !== 'undefined',
            webWorkers: typeof Worker !== 'undefined',
            webAssembly: typeof WebAssembly !== 'undefined',
            touchSupport: 'ontouchstart' in window || navigator.maxTouchPoints > 0
        };

    } catch (error) {
        console.warn('Error generating fingerprint:', error);
    }

    return fingerprint;
}

/**
 * Generate canvas fingerprint
 * @returns {Promise<string>} Canvas fingerprint hash
 */
function getCanvasFingerprint() {
    return new Promise((resolve) => {
        try {
            const canvas = document.createElement('canvas');
            const ctx = canvas.getContext('2d');

            canvas.width = 200;
            canvas.height = 50;

            // Draw fingerprinting pattern
            ctx.textBaseline = 'top';
            ctx.font = '14px Arial';
            ctx.fillStyle = '#f60';
            ctx.fillRect(125, 1, 62, 20);
            ctx.fillStyle = '#069';
            ctx.fillText('Hello, world! 🌍', 2, 15);
            ctx.fillStyle = 'rgba(102, 204, 0, 0.7)';
            ctx.fillText('Hello, world! 🌍', 4, 17);

            // Get canvas data
            const dataURL = canvas.toDataURL();
            resolve(dataURL);
        } catch (error) {
            resolve('canvas_error');
        }
    });
}

/**
 * Generate WebGL fingerprint
 * @returns {Object} WebGL information
 */
function getWebGLFingerprint() {
    try {
        const canvas = document.createElement('canvas');
        const gl = canvas.getContext('webgl') || canvas.getContext('experimental-webgl');

        if (!gl) {
            return { vendor: 'no_webgl', renderer: 'no_webgl' };
        }

        const debugInfo = gl.getExtension('WEBGL_debug_renderer_info');

        return {
            vendor: debugInfo ? gl.getParameter(debugInfo.UNMASKED_VENDOR_WEBGL) : 'unknown',
            renderer: debugInfo ? gl.getParameter(debugInfo.UNMASKED_RENDERER_WEBGL) : 'unknown',
            version: gl.getParameter(gl.VERSION),
            shadingLanguageVersion: gl.getParameter(gl.SHADING_LANGUAGE_VERSION)
        };
    } catch (error) {
        return { vendor: 'webgl_error', renderer: 'webgl_error' };
    }
}

/**
 * Detect available fonts (simplified approach)
 * @returns {Promise<Array>} List of detected fonts
 */
function detectFonts() {
    return new Promise((resolve) => {
        try {
            const baseFonts = ['monospace', 'sans-serif', 'serif'];
            const testFonts = [
                'Arial', 'Helvetica', 'Times New Roman', 'Courier New', 'Verdana',
                'Georgia', 'Palatino', 'Garamond', 'Bookman', 'Trebuchet MS',
                'Impact', 'Comic Sans MS', 'Tahoma', 'Lucida Console'
            ];

            const detectedFonts = [];
            const testString = 'mmmmmmmmmmlli';
            const testSize = '72px';

            // Create test element
            const testElement = document.createElement('div');
            testElement.style.position = 'absolute';
            testElement.style.left = '-9999px';
            testElement.style.fontSize = testSize;
            testElement.innerHTML = testString;
            document.body.appendChild(testElement);

            // Get baseline widths
            const baselineWidths = {};
            baseFonts.forEach(font => {
                testElement.style.fontFamily = font;
                baselineWidths[font] = testElement.offsetWidth;
            });

            // Test each font
            testFonts.forEach(font => {
                baseFonts.forEach(baseFont => {
                    testElement.style.fontFamily = `"${font}", ${baseFont}`;
                    if (testElement.offsetWidth !== baselineWidths[baseFont]) {
                        if (!detectedFonts.includes(font)) {
                            detectedFonts.push(font);
                        }
                    }
                });
            });

            document.body.removeChild(testElement);
            resolve(detectedFonts);
        } catch (error) {
            resolve([]);
        }
    });
}

/**
 * Create session data object for UMS API
 * @returns {Promise<Object>} Session creation object
 */
export async function createSessionData() {
    const fingerprint = await generateFingerprint();

    return {
        fingerprint: {
            components: fingerprint,
            browser_name: getBrowserName(fingerprint.userAgent),
            os_name: getOSName(fingerprint.userAgent),
            device_type: getDeviceType(fingerprint.userAgent),
            screen_resolution: fingerprint.screenResolution ?
                `${fingerprint.screenResolution[0]}x${fingerprint.screenResolution[1]}` : null,
            timezone: fingerprint.timezone,
            language: fingerprint.language
        },
        device_info: {
            screen_width: fingerprint.screenResolution ? fingerprint.screenResolution[0] : null,
            screen_height: fingerprint.screenResolution ? fingerprint.screenResolution[1] : null,
            pixel_ratio: fingerprint.pixelRatio,
            hardware_concurrency: fingerprint.hardwareConcurrency,
            device_memory: fingerprint.deviceMemory,
            touch_support: fingerprint.features?.touchSupport || false
        }
    };
}

/**
 * Extract browser name from user agent
 * @param {string} userAgent - User agent string
 * @returns {string} Browser name
 */
function getBrowserName(userAgent) {
    if (userAgent.includes('Chrome') && !userAgent.includes('Edge')) return 'Chrome';
    if (userAgent.includes('Firefox')) return 'Firefox';
    if (userAgent.includes('Safari') && !userAgent.includes('Chrome')) return 'Safari';
    if (userAgent.includes('Edge')) return 'Edge';
    if (userAgent.includes('Opera')) return 'Opera';
    return 'Unknown';
}

/**
 * Extract OS name from user agent
 * @param {string} userAgent - User agent string
 * @returns {string} OS name
 */
function getOSName(userAgent) {
    if (userAgent.includes('Windows')) return 'Windows';
    if (userAgent.includes('Mac OS') || userAgent.includes('macOS')) return 'macOS';
    if (userAgent.includes('Linux') && !userAgent.includes('Android')) return 'Linux';
    if (userAgent.includes('Android')) return 'Android';
    if (userAgent.includes('iPhone') || userAgent.includes('iPad')) return 'iOS';
    return 'Unknown';
}

/**
 * Determine device type from user agent
 * @param {string} userAgent - User agent string
 * @returns {string} Device type
 */
function getDeviceType(userAgent) {
    if (/Mobile|Android/.test(userAgent)) return 'mobile';
    if (/Tablet|iPad/.test(userAgent)) return 'tablet';
    return 'desktop';
}

/**
 * Store session token in local storage
 * @param {string} token - Session token
 */
export function storeSessionToken(token) {
    try {
        localStorage.setItem('ums_session_token', token);
    } catch (error) {
        console.warn('Could not store session token:', error);
    }
}

/**
 * Retrieve session token from local storage
 * @returns {string|null} Session token or null
 */
export function getSessionToken() {
    try {
        return localStorage.getItem('ums_session_token');
    } catch (error) {
        console.warn('Could not retrieve session token:', error);
        return null;
    }
}

/**
 * Clear session token from local storage
 */
export function clearSessionToken() {
    try {
        localStorage.removeItem('ums_session_token');
    } catch (error) {
        console.warn('Could not clear session token:', error);
    }
}

File: ums/utils/fingerprint.py
"""
Browser fingerprinting utilities for user recognition

This module provides functions to generate and normalize browser fingerprints
for identifying returning anonymous users.
"""

import hashlib
import json
from typing import Dict, Any, Optional


def normalize_fingerprint(raw_fingerprint: Dict[str, Any]) -> Dict[str, Any]:
    """
    Normalize and clean fingerprint data to improve matching accuracy

    Args:
        raw_fingerprint: Raw fingerprint data from browser

    Returns:
        Normalized fingerprint data
    """
    normalized = {}

    # Browser information
    if 'userAgent' in raw_fingerprint:
        normalized['user_agent'] = raw_fingerprint['userAgent']

    if 'language' in raw_fingerprint:
        normalized['language'] = raw_fingerprint['language']

    if 'languages' in raw_fingerprint:
        normalized['languages'] = sorted(raw_fingerprint['languages'])

    # Screen information (rounded for privacy)
    if 'screenResolution' in raw_fingerprint:
        width, height = raw_fingerprint['screenResolution']
        # Round to nearest 100 for privacy
        normalized['screen_width'] = round(width / 100) * 100
        normalized['screen_height'] = round(height / 100) * 100

    if 'colorDepth' in raw_fingerprint:
        normalized['color_depth'] = raw_fingerprint['colorDepth']

    if 'pixelRatio' in raw_fingerprint:
        # Round to 1 decimal place
        normalized['pixel_ratio'] = round(raw_fingerprint['pixelRatio'], 1)

    # Timezone
    if 'timezone' in raw_fingerprint:
        normalized['timezone'] = raw_fingerprint['timezone']

    # Platform information
    if 'platform' in raw_fingerprint:
        normalized['platform'] = raw_fingerprint['platform']

    # Hardware information (if available)
    if 'hardwareConcurrency' in raw_fingerprint:
        normalized['cpu_cores'] = raw_fingerprint['hardwareConcurrency']

    if 'deviceMemory' in raw_fingerprint:
        normalized['device_memory'] = raw_fingerprint['deviceMemory']

    # Canvas fingerprint (if available)
    if 'canvas' in raw_fingerprint:
        normalized['canvas'] = raw_fingerprint['canvas']

    # WebGL information (if available)
    if 'webgl' in raw_fingerprint:
        webgl = raw_fingerprint['webgl']
        if isinstance(webgl, dict):
            normalized['webgl_vendor'] = webgl.get('vendor', '')
            normalized['webgl_renderer'] = webgl.get('renderer', '')

    # Fonts (if available)
    if 'fonts' in raw_fingerprint:
        # Sort fonts for consistent hashing
        normalized['fonts'] = sorted(raw_fingerprint['fonts'])

    # Plugins (if available)
    if 'plugins' in raw_fingerprint:
        # Sort plugins for consistent hashing
        normalized['plugins'] = sorted(raw_fingerprint['plugins'])

    return normalized


def generate_fingerprint(raw_fingerprint: Dict[str, Any]) -> str:
    """
    Generate a stable hash from normalized fingerprint data

    Args:
        raw_fingerprint: Raw fingerprint data from browser

    Returns:
        SHA256 hash of the normalized fingerprint
    """
    normalized = normalize_fingerprint(raw_fingerprint)

    # Convert to JSON string with sorted keys for consistent hashing
    fingerprint_string = json.dumps(normalized, sort_keys=True, separators=(',', ':'))

    # Generate SHA256 hash
    fingerprint_hash = hashlib.sha256(fingerprint_string.encode('utf-8')).hexdigest()

    return fingerprint_hash


def calculate_fingerprint_strength(raw_fingerprint: Dict[str, Any]) -> int:
    """
    Calculate the strength/uniqueness of a fingerprint

    Args:
        raw_fingerprint: Raw fingerprint data from browser

    Returns:
        Strength score from 0-100 (higher = more unique)
    """
    score = 0
    components = 0

    # Basic browser info (10 points)
    if 'userAgent' in raw_fingerprint:
        score += 10
        components += 1

    # Language info (5 points)
    if 'language' in raw_fingerprint:
        score += 5
        components += 1

    # Screen resolution (15 points - highly identifying)
    if 'screenResolution' in raw_fingerprint:
        score += 15
        components += 1

    # Timezone (10 points - fairly identifying)
    if 'timezone' in raw_fingerprint:
        score += 10
        components += 1

    # Hardware info (10 points each)
    if 'hardwareConcurrency' in raw_fingerprint:
        score += 10
        components += 1

    if 'deviceMemory' in raw_fingerprint:
        score += 10
        components += 1

    # Canvas fingerprint (20 points - very identifying)
    if 'canvas' in raw_fingerprint:
        score += 20
        components += 1

    # WebGL info (15 points - fairly identifying)
    if 'webgl' in raw_fingerprint:
        score += 15
        components += 1

    # Fonts (10 points)
    if 'fonts' in raw_fingerprint:
        font_count = len(raw_fingerprint['fonts'])
        score += min(font_count // 10, 10)  # Up to 10 points based on font count
        components += 1

    # Plugins (5 points)
    if 'plugins' in raw_fingerprint:
        score += 5
        components += 1

    return min(score, 100)


def compare_fingerprints(fp1_hash: str, fp2_hash: str) -> float:
    """
    Compare two fingerprint hashes for similarity

    Args:
        fp1_hash: First fingerprint hash
        fp2_hash: Second fingerprint hash

    Returns:
        Similarity score from 0.0-1.0 (1.0 = identical)
    """
    if fp1_hash == fp2_hash:
        return 1.0

    # For hash-based fingerprints, they're either identical or completely different
    # In the future, we could implement fuzzy matching for more sophisticated comparison
    return 0.0


def extract_device_info(raw_fingerprint: Dict[str, Any]) -> Dict[str, str]:
    """
    Extract device information from fingerprint data

    Args:
        raw_fingerprint: Raw fingerprint data from browser

    Returns:
        Dictionary with device information
    """
    device_info = {}

    user_agent = raw_fingerprint.get('userAgent', '')

    # Extract browser name and version
    if 'Chrome' in user_agent:
        device_info['browser'] = 'Chrome'
    elif 'Firefox' in user_agent:
        device_info['browser'] = 'Firefox'
    elif 'Safari' in user_agent:
        device_info['browser'] = 'Safari'
    elif 'Edge' in user_agent:
        device_info['browser'] = 'Edge'
    else:
        device_info['browser'] = 'Unknown'

    # Extract OS
    if 'Windows' in user_agent:
        device_info['os'] = 'Windows'
    elif 'Mac OS' in user_agent or 'macOS' in user_agent:
        device_info['os'] = 'macOS'
    elif 'Linux' in user_agent:
        device_info['os'] = 'Linux'
    elif 'Android' in user_agent:
        device_info['os'] = 'Android'
    elif 'iOS' in user_agent:
        device_info['os'] = 'iOS'
    else:
        device_info['os'] = 'Unknown'

    # Determine device type
    if 'Mobile' in user_agent or 'Android' in user_agent:
        device_info['device_type'] = 'mobile'
    elif 'Tablet' in user_agent or 'iPad' in user_agent:
        device_info['device_type'] = 'tablet'
    else:
        device_info['device_type'] = 'desktop'

    # Extract screen resolution
    if 'screenResolution' in raw_fingerprint:
        width, height = raw_fingerprint['screenResolution']
        device_info['screen_resolution'] = f"{width}x{height}"

    # Extract timezone
    if 'timezone' in raw_fingerprint:
        device_info['timezone'] = raw_fingerprint['timezone']

    # Extract language
    if 'language' in raw_fingerprint:
        device_info['language'] = raw_fingerprint['language']

    return device_info